### DISCUSSION TITLE
PicoOTA update from SD Storage?

### DISCUSSION THREAD

**Discussion by Jordessey** [earlephilhower/arduino-pico#2354] [Q&A]
Hey Smart People,

I have an idea but not sure if its supported/how to put into action.
with the PicoOTA from file, is it possible to pull the file from an SD Card through the SD.h?

Thanks :-)

**Comment by earlephilhower** [comment]
It would be possible with much work and probably take more space than the current LittleFS-based one (although w/o gzip compression support it might be close).

OTA runs with the bare SDK and as little running as possible, so no SD.h.  You'd have to get SDFat or FatFS (I'd go w/FatFS) running bare metal and with as little of the SDK as possible.  SDFat has SD protocol built in, FatFS doesn't (but I thing Elm-Chan does have some implementation).  You'd also need some way of imprinting the SPI settings in the bootloader because every board has its own SPI config.

If you're really interested I would suggest first building a copy-to-RAM executable using the bare SDK that does this.  Once that's running you can start chopping out whatever you can to shrink it.  See the `ota/` directory for our own implementation here.
